tests/linearizability/model.go

```diff
@@ -209,5 +231,51 @@ func stepDelete(state EtcdState, request EtcdRequest, response EtcdResponse) (bo
 
 	state.Value = ""
 	state.LastRevision = response.Revision
+	state.FailedWrite = nil
+	return true, state
+}
+
+func stepTxn(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
+	if response.Err != nil {
+		state.FailedWrite = &request
+		return true, state
+	}
+	// revision should never decrease
+	if response.Revision < state.LastRevision {
+		return false, state
+	}
+	valueMatch := state.Value == request.TxnExpectData
+
+	if state.FailedWrite == nil {
+		// transaction should succeed if value matches
+		if response.TxnSucceeded != valueMatch {
+			return false, state
+		}
+
+		// if transaction succeeded, revision must increase
+		if response.TxnSucceeded && response.Revision != state.LastRevision+1 {
+			return false, state
+		}
+		// if transaction failed, revision stays the same
+		if !response.TxnSucceeded && response.Revision != state.LastRevision {
+			return false, state
+		}
+	} else {
+		// if value matches, revision must increase as either transaction succeeded or failed request was persisted.
+		if valueMatch && response.Revision < state.LastRevision+1 {
+			return false, state
+		}
+
+		// if txn succeeded (+1 rev) without matching key, means that fail request was persisted (+1 rev).
+		if response.TxnSucceeded && !valueMatch && response.Revision < state.LastRevision+2 {
+			return false, state
+		}
```

Comment:
Note that we shouldn't depend on `valueMatch` at all.
```suggestion
		// See https://github.com/etcd-io/etcd/pull/14902#issuecomment-1340041130
		if response.TxnSucceeded && response.Revision < state.LastRevision+1 {
			return false, state
		}
```